### PR TITLE
Expand VisitMachineValues to cover more pointers in the interpreter

### PR DESCRIPTION
**Pull request by saethlin**
Follow-on to https://github.com/rust-lang/miri/pull/2559

This is making me want to write a proc macro :thinking: 

r? @RalfJung

### CODE REVIEW

File: src/machine.rs [hunk 1]
```diff
@@ -591,12 +601,26 @@ impl<'mir, 'tcx> MiriMachine<'mir, 'tcx> {
...
+        }
+        if let Some(argv) = argv {
+            visit(&Operand::Indirect(*argv));
+        }
+        if let Some(cmd_line) = cmd_line {
+            visit(&Operand::Indirect(*cmd_line));
+        }
+        for ptr in extern_statics.values().copied() {
+            let ptr: Pointer<Option<Provenance>> = ptr.into();
+            visit(&Operand::Indirect(MemPlace::from_ptr(ptr)));
```
Comment (saethlin) [inline src/machine.rs:622]: This is pretty ugly. Is there a better way to do this? Or should we adjust the interface of the visitor so we can just pass a pointer to it directly?

File: src/tag_gc.rs [hunk 2]
```diff
@@ -71,6 +71,22 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: MiriInterpCxExt<'mir, 'tcx> {
...
+
+                if let Some(store_buffers) = alloc.extra.weak_memory.as_ref() {
+                    store_buffers.iter(|val| {
+                        if let Scalar::Ptr(ptr, _) = val {
+                            if let Provenance::Concrete { sb, .. } = ptr.provenance {
+                                tags.insert(sb);
+                            }
+                        }
+                    });
+                }
```
Comment (saethlin) [inline src/tag_gc.rs:89]: I think this is exactly the kind of code you were trying to get rid of with the visitor, but I'm really not sure what to do here because we're implementing the visitor.

File: src/concurrency/thread.rs [hunk 3]
```diff
@@ -181,6 +181,41 @@ impl<'mir, 'tcx> Thread<'mir, 'tcx> {
     }
 }
 
+impl VisitMachineValues for Thread<'_, '_> {
+    fn visit_machine_values(&self, visit: &mut impl FnMut(&Operand<Provenance>)) {
+        let Thread { panic_payload, last_error, stack, .. } = self;
```
Comment (RalfJung) [inline src/concurrency/thread.rs:186]: My plan was to never use `..` here (since then we might add more fields and forget about visiting them) and instead explicitly do `field: _` for fields that do not visiting. RA can help with adding all those field names.

File: src/concurrency/thread.rs [hunk 4]
```diff
@@ -253,6 +288,22 @@ impl<'mir, 'tcx> Default for ThreadManager<'mir, 'tcx> {
...
+        let ThreadManager { threads, thread_local_alloc_ids, .. } = self;
+
+        for thread in threads {
+            thread.visit_machine_values(visit);
+        }
+        for ptr in thread_local_alloc_ids.borrow().values().copied() {
+            let ptr: Pointer<Option<Provenance>> = ptr.into();
+            visit(&Operand::Indirect(MemPlace::from_ptr(ptr)));
+        }
+        // FIXME: Do we need to do something for TimeoutCallback? That's a Box<dyn>, not sure what
```
Comment (RalfJung) [inline src/concurrency/thread.rs:302]: Oh, good catch. That's tricky to handle -- if we let callbacks capture things (and I assume they do capture something) then we either have to declare that they must not capture any machine values (which is tricky to check) or we have to replace the closures by an explicit struct + trait.

File: src/machine.rs (hunk 1 above)
Comment (RalfJung) [reply src/machine.rs:632]: We could have more `From` instances to convert pointers to MemPlaces or so... :shrug:

File: src/concurrency/weak_memory.rs [hunk 5]
```diff
@@ -108,6 +108,19 @@ pub struct StoreBufferAlloc {
     store_buffers: RefCell<RangeObjectMap<StoreBuffer>>,
 }
 
+impl StoreBufferAlloc {
+    pub fn iter(&self, mut visitor: impl FnMut(&Scalar<Provenance>)) {
```
Comment (RalfJung) [inline src/concurrency/weak_memory.rs:112]: Why does this need to be `pub`? I'd expect a visitor impl somewhere inside `src/concurrency`.

File: src/stacked_borrows/mod.rs [hunk 6]
```diff
@@ -79,7 +79,7 @@ pub struct Stacks {
     /// Stores past operations on this allocation
     history: AllocHistory,
     /// The set of tags that have been exposed inside this allocation.
-    exposed_tags: FxHashSet<SbTag>,
+    pub exposed_tags: FxHashSet<SbTag>,
```
Comment (RalfJung) [inline src/stacked_borrows/mod.rs:82]: Why does this need to be `pub`? I'd expect a visitor impl somewhere inside `src/stacked_borrows`.

File: src/tag_gc.rs (hunk 2 above)
Comment (RalfJung) [reply src/tag_gc.rs:89]: Hm, yeah allocations are a bit tricky to handle together with the rest of this.

We could have a `VisitProvenance` trait and implement that for `Allocation` and `AllocExtra` and recursively their fields, like we do for machine values?

File: src/stacked_borrows/stack.rs [hunk 7]
```diff
@@ -43,8 +43,11 @@ impl Stack {
     pub fn retain(&mut self, tags: &FxHashSet<SbTag>) {
         let mut first_removed = None;
 
-        let mut read_idx = 1;
-        let mut write_idx = 1;
+        // For stacks with a known bottom, we never consider removing the bottom-most tag, because
+        // that is the base tag which exists whether or not there are any pointers to the
+        // allocation.
+        let mut read_idx = usize::from(self.unknown_bottom.is_none());
```
Comment (RalfJung) [inline src/stacked_borrows/stack.rs:49]: The code is double-negated when compared with the text. I think `if self.unknown_bottom.is_some() { 0 } else { 1 }` would be more clear.

File: src/concurrency/weak_memory.rs (hunk 5 above)
Comment (saethlin) [reply src/concurrency/weak_memory.rs:112]: I don't know how to do this with a visitor impl, because this is "visited" in the implementation of the visitor. This is a part of `AllocExtra`, not a member of `Machine`.

File: src/concurrency/thread.rs (hunk 3 above)
Comment (saethlin) [reply src/concurrency/thread.rs:186]: `Thread` has a private field so I can add `field: _` for all the public fields but we can't get rid of the `..`.

Comment (RalfJung) [reply src/concurrency/thread.rs:186]: This is the file defining `Thread` so you should be able to access all private fields.

File: src/concurrency/weak_memory.rs (hunk 5 above)
Comment (RalfJung) [reply src/concurrency/weak_memory.rs:112]: > This is a part of AllocExtra, not a member of Machine.

So what?
`machine::AllocExtra` should also implement the trait so that all the per-allocation state is being visited.

File: src/concurrency/thread.rs (hunk 3 above)
Comment (saethlin) [reply src/concurrency/thread.rs:186]: Oh I meant `Frame` :facepalm:

File: src/concurrency/weak_memory.rs (hunk 5 above)
Comment (saethlin) [reply src/concurrency/weak_memory.rs:112]: As far as I understand, we call `this.machine.vist_all_machine_values` then from the `Machine` we recurse down to all the parts of `Machine`. I don't know how to recurse down into the allocations from the `Machine`.

File: src/concurrency/thread.rs (hunk 4 above)
Comment (saethlin) [reply src/concurrency/thread.rs:318]: I don't know if what I implemented is exactly what you had in mind, but I ended up with a surprised lifetime bound. I suspect it comes from the fact that I'm not using `FnOnce`? I suppose it's possible I also got a `'mir` and `'tcx` backwards but I've been over the code a bunch of times now :sleeping:

File: src/concurrency/thread.rs [hunk 8]
```diff
@@ -32,9 +32,9 @@ pub enum SchedulingAction {
 
 /// Timeout callbacks can be created by synchronization primitives to tell the
 /// scheduler that they should be called once some period of time passes.
-type TimeoutCallback<'mir, 'tcx> = Box<
-    dyn FnOnce(&mut InterpCx<'mir, 'tcx, MiriMachine<'mir, 'tcx>>) -> InterpResult<'tcx> + 'tcx,
->;
+pub trait TimeoutCallback<'mir, 'tcx>: VisitMachineValues + 'tcx {
+    fn call(&self, ecx: &mut InterpCx<'mir, 'tcx, MiriMachine<'mir, 'tcx>>) -> InterpResult<'tcx>;
+}
```
Comment (oli-obk) [inline src/concurrency/thread.rs:37]: You need to remove the `+'tcx` from the super trait list and add it only to the `Box<dyn Trait + 'tcx>`, otherwise you get an implication that `'mir: 'tcx` due to the `'mir` on the trait

Comment (saethlin) [reply src/concurrency/thread.rs]: Hunh. I'm sure that will seem intuitive to me later.

File: src/concurrency/thread.rs (hunk 4 above)
Comment (saethlin) [reply src/concurrency/thread.rs]: Oli got me straightened out here, still not entirely sure of the implementation.

File: src/concurrency/thread.rs (hunk 3 above)
Comment (RalfJung) [reply src/concurrency/thread.rs]: Yeah okay `Frame` will need a `..` with a comment explaining why.

File: src/concurrency/thread.rs (hunk 4 above)
Comment (RalfJung) [reply src/concurrency/thread.rs]: Yes that looks like the struct+trait solution. :)

File: src/shims/unix/linux/sync.rs [hunk 9]
```diff
@@ -193,14 +193,7 @@ pub fn futex<'tcx>(
...
                         timeout_time,
-                        Box::new(move |this| {
-                            this.unblock_thread(thread);
-                            this.futex_remove_waiter(addr_usize, thread);
-                            let etimedout = this.eval_libc("ETIMEDOUT")?;
-                            this.set_last_error(etimedout)?;
-                            this.write_scalar(Scalar::from_machine_isize(-1, this), &dest)?;
-                            Ok(())
-                        }),
+                        Box::new(Callback { thread, addr_usize, dest }),
```
Comment (RalfJung) [inline src/shims/unix/linux/sync.rs:196]: Can you define this struct inline in this function? I think we can and should keep the code together in one place, like it was before with the closure.

File: src/machine.rs [hunk 10]
```diff
@@ -590,13 +614,36 @@ impl<'mir, 'tcx> MiriMachine<'mir, 'tcx> {
...
+        let MiriMachine {
+            threads,
+            tls,
+            env_vars,
+            argc,
+            argv,
+            cmd_line,
+            extern_statics,
+            dir_handler,
+            ..
```
Comment (RalfJung) [inline src/machine.rs:627]: With RA it should be fairly easy to add all the missing `field: _` here?

When the next contributor adds a new field we need to be reminded of the visitor...

File: src/tag_gc.rs [hunk 11]
```diff
@@ -3,30 +3,120 @@ use rustc_data_structures::fx::FxHashSet;
 use crate::*;
 
 pub trait VisitMachineValues {
-    fn visit_machine_values(&self, visit: &mut impl FnMut(&Operand<Provenance>));
+    fn visit_machine_values(&self, visit: &mut ProvenanceVisitor);
+}
```
Comment (RalfJung) [inline src/tag_gc.rs:7]: Hm okay this is less general than what I started out with, but I guess for now visiting all provenance is sufficient.

But then it should probably be called `VistiProvenance`.
And why did you change the `visit` argument?

File: src/tag_gc.rs [hunk 12]
```diff
@@ -3,30 +3,120 @@ use rustc_data_structures::fx::FxHashSet;
...
+
+pub trait MachineValue {
```
Comment (RalfJung) [inline src/tag_gc.rs:9]: I don't understand the point of this trait. Why not to `impl VisitProvenance for Operand` etc?

File: src/tag_gc.rs [hunk 13]
```diff
@@ -3,30 +3,120 @@ use rustc_data_structures::fx::FxHashSet;
...
+    }
+}
+
+impl<T: MachineValue> MachineValue for &T {
+    fn visit_provenance(&self, tags: &mut FxHashSet<SbTag>) {
+        (**self).visit_provenance(tags);
+    }
+}
+
+impl MachineValue for Operand<Provenance> {
```
Comment (RalfJung) [inline src/tag_gc.rs:32]: Now that we have a bunch of stuff here, might be good to have a new file `machine_visitor.rs` or so.

File: src/shims/unix/sync.rs [hunk 14]
```diff
@@ -898,6 +885,39 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
     }
 }
 
+struct Callback<'tcx> {
```
Comment (RalfJung) [inline src/shims/unix/sync.rs:888]: This is also just used in one place so could be moved inside `pthread_cond_timedwait` -- or else it should get a descriptive name + docs.

File: src/stacked_borrows/mod.rs [hunk 15]
```diff
@@ -513,6 +513,14 @@ impl Stacks {
     }
 }
 
+impl VisitMachineValues for Stacks {
+    fn visit_machine_values(&self, visit: &mut ProvenanceVisitor) {
+        for tag in self.exposed_tags.iter().copied() {
```
Comment (RalfJung) [inline src/stacked_borrows/mod.rs:518]: I'm a bit surprised by the 3 `copied()` you added, given that visiting should work by-ref.

File: src/tag_gc.rs (hunk 11 above)
Comment (saethlin) [reply src/tag_gc.rs:7]: We need to visit a number of different things. It's possible to contort them all into an `Operand`, but that adds a lot of repetitive glue code at the call sites. If each use of the visitor can take a different type, we can do the dispatch inside this module, instead of spread across the codebase. So we need to shift the generic parameter to each call to the visitor, which means we need a new type with the generic parameter on the method, not `FnMut`.

Comment (RalfJung) [reply src/tag_gc.rs:7]: But it seems you solved this by completely re-structuring the visitor to (a) to longer be lazy (it now hard-codes the use of a hash map -- `ProvenanceVisitor` is a somewhat misleading name) and (b) hard-code that only provenance matters.

If we want to hard-code that only provenance matters (which I am fine with for now), we should have
```rust
pub trait VisitProvenance {
 fn visit_provenance(&self, visit: &mut impl FnMut(SbTag));
}
```
Then you can `impl VisitProvenance for {Operand,Immediate,Pointer}` and you can now do `x.visit_provenance(visit)` for `x ∈ {Operand,Immediate,Pointer}`.

If we don't want to hard-code hat only provenance matters we need something else, but your approach does not achieve that.

Comment (RalfJung) [reply src/tag_gc.rs:7]: That said, some places only have `SbTag`, not a full `Provenance`... not sure what the best way is to handle them. We could make the visitor focus only on `SbTag` I guess; then it should be called `VisitProvenanceTags` or so.

File: src/stacked_borrows/mod.rs (hunk 15 above)
Comment (saethlin) [reply src/stacked_borrows/mod.rs:518]: Refactoring churn.

File: src/tag_gc.rs (hunk 11 above)
Comment (RalfJung) [reply src/tag_gc.rs:7]: I pushed to this branch a version of these traits (or rather, just a single trait) demonstrating what I meant.

File: src/shims/panic.rs [hunk 16]
```diff
@@ -35,6 +35,14 @@ pub struct CatchUnwindData<'tcx> {
     ret: mir::BasicBlock,
 }
 
+impl VisitMachineValues for CatchUnwindData<'_> {
+    fn visit_machine_values(&self, visit: &mut ProvenanceVisitor) {
+        let CatchUnwindData { catch_fn, data, dest: _, ret: _ } = self;
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply src/shims/panic.rs]: This one was still missing...

**Comment by RalfJung** [comment]
I went over all the visitors and made sure the discarded parts with `_` have a type that doesn't need visit.

But yeah some kind of macro would be great.^^ Even a regular macro where one lists all the fields and that does the pattern expansion and calling the visitor. rustc used to have something like that but then later it got turned into a proper proc macro.

**Comment by RalfJung** [comment]
Ah, CI already merges with master. We can't rebase due to the subtree sync situation.

**Review by oli-obk — APPROVED** [review]
